### ISSUE TITLE
Add support for `${XDG_CONFIG_HOME}/containers/auth.json`

### ISSUE THREAD

**Issue by PhrozenByte** [containers/podman#12706]
/kind feature

**Description**

Podman currently expects auth files in `${XDG_RUNTIME_DIR}/containers/auth.json` and, for Docker compatibility, in `$HOME/.docker/config.json`. Alternatively one can pass the `--authfile` option to most Podman commands, or set the `REGISTRY_AUTH_FILE` environment variable, to use arbitrary paths.

I'd like to suggest that Podman additionally checks in `${XDG_CONFIG_HOME}/containers/auth.json` for an auth file, if authentication is required and no explicit path was given.

To enforce some security, Podman could reject `${XDG_CONFIG_HOME}/containers/auth.json` if permissions are too loose (i.e. the file must be readable by the executing user only; think about e.g. SSH clients enforcing secure file permissions for SSH keys).

**Reasoning**

`${XDG_RUNTIME_DIR}/containers/auth.json` is the right choice for `podman login`. On most systems `${XDG_RUNTIME_DIR}` is a `tmpfs` and storing `podman login` credentials there is just what one would expect from a non-persistent login.

However, if one tries to run services from images managed in private registries, one needs a persistent credentials storage (think about e.g. FCOS). Due to `${XDG_RUNTIME_DIR}`'s non-persistent nature, Podman lacks a native way to provide credentials persistently besides the Docker fallback `$HOME/.docker/config.json`. For most commands (like `podman run`) we can indeed simply pass `--authfile` on the command line (resp. the Systemd unit), however, for some features, like `podman auto-update`, we don't have such an easy solution.

Adding `${XDG_CONFIG_HOME}/containers/auth.json` to the "credentials search path" seems like a simple and reasonable enhancement.

**Comment by mtrmac** [comment]
> however, for some features, like `podman auto-update`, we don't have such an easy solution.

For _non-interactive automated use_, it seems better to me to just provide the `--authfile` option to all commands.

In such cases, users should not need to run `podman login` inside the non-interactive automated script, they can just provide an `auth.json` file as a prepared artifact as part of the input to the non-interactive operation. So it seems natural to me to have that as an explicit input to the automated operation, rather than share state with any other operations that might concurrently use the same user account: it would be more explicit and better-isolated.

---

If we add a `$HOME/…auth.json`, my first question is “where does `podman login` write afterwards”? At least if `/run/user/…` is available on a tmpfs, writing to $HOME would be a decrease in security.

Of course that question has several plausible answers that would _work_, but I’m worried that none of them are obviously _intuitive_ to users who just read tutorials on the internet.

**Comment by matejvasek** [comment]
What about providing some envvar (e.g. `PODMAN_AUTHFILE`) so the `--authfile` flag doesn't have to be repeated?

**Comment by PhrozenByte** [comment]
> If we add a `$HOME/…auth.json`, my first question is “where does `podman login` write afterwards”? At least if `/run/user/…` is available on a tmpfs, writing to $HOME would be a decrease in security.

This issue just aims at adding `${XDG_CONFIG_HOME}/containers/auth.json` to Podman's credentials search path (i.e. reading credentials, not writing them).

> In such cases, users should not need to run `podman login` inside the non-interactive automated script, they can just provide an `auth.json` file as a prepared artifact as part of the input to the non-interactive operation. So it seems natural to me to have that as an explicit input to the automated operation, rather than share state with any other operations that might concurrently use the same user account: it would be more explicit and better-isolated.

I'd say that this is off-topic. As far as I get you right, you basically suggest that `podman auto-update` should no longer check `${XDG_RUNTIME_DIR}/containers/auth.json`, because you wouldn't expect the (putative) non-interactive `podman auto-update` to be affected by the (putative) interactive `podman login`. I get your point, but I don't really see how this is related to adding `${XDG_CONFIG_HOME}/containers/auth.json` as persistent (non-interactive) credentials storage as an alternative to Docker's `$HOME/.docker/config.json`.

If you want to see this changed you should probably create a new issue for this.

**Comment by mtrmac** [comment]
No, I’m not suggesting _dropping_ the current path from `podman auto-update`.

But as one of the people likely to be either implementing the proposal or reviewing an implementation, I’m documenting how I think about this issue, to set expectations and to steer the conversation. It’s easy enough to write the code that implements this extra path, but I think it’s more useful to discuss whether we _want_ that, before taking the effort to implement it. That conversation is going to happen either way, so we might as well do it first.

**Comment by mtrmac** [comment]
> What about providing some envvar (e.g. `PODMAN_AUTHFILE`) so the `--authfile` flag doesn't have to be repeated?
> EDIT: sorry, now I see such an envvar already exists.

Yes, `REGISTRY_AUTH_FILE`.

**Comment by PhrozenByte** [comment]
> No, I’m not suggesting _dropping_ the current path from `podman auto-update`.

Then I unfortunately understand even less what you're trying to say and how it concerns this suggestion. Please elaborate further, as it's kinda hard to discuss it otherwise.

**Comment by mtrmac** [comment]
I’d, right now, prefer adding an `--authfile` option to Podman subcommands that lack it (which seems clearly useful and uncontroversial) to adding another `auth.json` location (which might be a useful feature without downsides, or an usability problem, and it’s not immediately obvious which one it is).

**Comment by PhrozenByte** [comment]
I guess that you got the issue wrong: This is not about `podman auto-update` and others lacking `--authfile` (see https://docs.podman.io/en/latest/markdown/podman-auto-update.1.html#authfile-path), but the fact that it can't be used properly.

Keep rootless containers in mind: With rootless containers one needs a `podman-auto-update.service` per user. Since auth files are user-specific for a reason, every service requires a different path for `--authfile` (like `/home/user/.config/containers/auth.json`). Thus we can't just enable a shared user service, but create unique user services per user. Using `--authfile` is furthermore semantically wrong. Authentication is a matter of the container to update, not of an overall service like `podman auto-update`. Right now one is required to update `podman-auto-update.service` if a container that requires authentication is added, or if all containers that require authentication are removed. Podman already supports persistent authentication with `$HOME/.docker/config.json`, it just lacks a Podman-native solution.

> which might be a useful feature without downsides, or a usability problem, and it’s not immediately obvious which one it is

Please elaborate on how this could be a "usability problem".

**Comment by mtrmac** [comment]
https://github.com/containers/podman/issues/12706#issuecomment-1026077760

**Comment by PhrozenByte** [comment]
I feel like this being very rude and unprofessional behaviour of yours @mtrmac, as you're unwilling to elaborate what you think the issue might be :-1: As explained earlier, this issue is not about changing the behaviour of `podman login`.

**Comment by rhatdan** [comment]
Why not set 
`export REGISTRY_AUTH_FILE=$XDG_CONFIG_HOME/containers/auth.json`

In your environment?

**Comment by mtrmac** [comment]
> this issue is not about changing the behaviour of `podman login`.

If we start reading a file, it automatically raises a question of how is that file going to be written, then. The whole set of use cases must make sense as a whole.

**Comment by PhrozenByte** [comment]
> Why not set
> `export REGISTRY_AUTH_FILE=$XDG_CONFIG_HOME/containers/auth.json`
> In your environment?

This is indeed the solution I chose shortly after opening this issue. However, I don't feel like that creating a `/etc/profile.d/containers_auth.sh` with

```sh
# /etc/profile.d/containers_auth.sh

if [ -f "${XDG_CONFIG_HOME:-$HOME/.config}/containers/auth.json" ]; then
    export REGISTRY_AUTH_FILE="${XDG_CONFIG_HOME:-$HOME/.config}/containers/auth.json"
fi
```

is a durable solution, it rather is a workaround. This is an issue that any Podman user that tries to setup auto updates for rootless containers with private registries faces. Right now Podman provides no native solution. Setting `REGISTRY_AUTH_FILE` as global environment variable works, but is changing the system's global environment really the officially recommended solution to this - I'd say at least not uncommon - use case?

If this is true it should be documented somewhere and this issue can be closed.

> If we start reading a file, it automatically raises a question of how is that file going to be written, then. The whole set of use cases must make sense as a whole.

Thank you for your elaboration. This is a valid point we should discuss further.

The need for persistent credential storage primarily arises from systems that are highly automated, from deployment to operations (you called it "non-interactive automated use") - like FCOS. Considering this I totally agree to the following you said earlier:

> In such cases, users should not need to run `podman login` inside the non-interactive automated script, they can just provide an `auth.json` file as a prepared artifact as part of the input to the non-interactive operation.

On such systems `${XDG_CONFIG_HOME}/containers/auth.json` is created only once on deployment. In the case of FCOS one would simply include it in their Ignition config (via Butane or OpenShift). So this issue won't arise on these systems, one simply adds a previously created `${XDG_CONFIG_HOME}/containers/auth.json` to the deployment and `podman auto-update` will pick it up. Right now this is neither possible with `${XDG_RUNTIME_DIR}/containers/auth.json`, nor with `--authfile` (at least not without creating a unique `podman-auto-update.service` user service per user, see https://github.com/containers/podman/issues/12706#issuecomment-1026706302).

Personally I'd leave it this way, there's no need to interactively write `${XDG_CONFIG_HOME}/containers/auth.json`, because I agree to what you said earlier: `${XDG_CONFIG_HOME}/containers/auth.json` is for non-interactive automated use. `podman login` is for interactive use and behaves exactly as expected (see original comment).

**Comment by rhatdan** [comment]
I don't think we will move forward on this, since this can be handled via scripts, I am going to close.
